### PR TITLE
Cache values in dependent project finder

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Removes like 60% of a lightbulb scenario:

image

image

Discovered when trying to figure out why a lightbulb was taking several seconds to load.

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs [hunk 1]
```diff
@@ -326,16 +374,25 @@ private static bool HasReferenceToAssembly(Project project, string assemblyName,
...
+            CreateCompilation(project, ref compilation);
+
+            if (compilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol { Name: string metadataAssemblyName })
+            {
+                using (await s_metadataIdToAssemblyNameGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    s_metadataIdToAssemblyName.TryAdd(metadataId, metadataAssemblyName);
+                    if (metadataAssemblyName == assemblyName)
+                        return true;
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs:393]:
```suggestion
```

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs [hunk 2]
```diff
@@ -307,14 +352,17 @@ private static bool HasReferenceToAssembly(Project project, string assemblyName,
             if (metadataId is null)
                 continue;
 
-            if (!s_metadataIdToAssemblyName.TryGetValue(metadataId, out var name))
+            using (await s_metadataIdToAssemblyNameGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
             {
-                uncomputedReferences.Add((peReference, metadataId));
-                continue;
-            }
+                if (!s_metadataIdToAssemblyName.TryGetValue(metadataId, out var name))
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs:357]: >if (!s_metadataIdToAssemblyName.TryGetValue(metadataId, out var name)) [](http://example.com/codeflow?start=16&length=70)

might be nice to match below:

if (s_metadataIdToAssemblyName.TryGetValue(metadataId, out var name) && name == assemblyName)
    return true;

and then move the uncomputedReferences outside the gate

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs]: done.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs [hunk 3]
```diff
@@ -128,24 +137,59 @@ internal static partial class DependentProjectsFinder
...
+        }
+
+        // Compute if not in cache.
+        dependentProjects = await ComputeDependentProjectsWorkerAsync(
+            solution, symbolOrigination, visibility, cancellationToken).ConfigureAwait(false);
+
+        // Try to add to cache, returning existing value if another thread already added it.
+        using (await s_solutionToDependentProjectMapGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (dictionary.TryGetValue((symbolOrigination.assembly, symbolOrigination.sourceProject, visibility), out dependentProjects))
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs]: Doesn't the GetOrAdd do the TryGetValue already?

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs [hunk 4]
```diff
@@ -128,24 +137,59 @@ internal static partial class DependentProjectsFinder
...
+                    solution, symbolOrigination, dependentProjects, cancellationToken).ConfigureAwait(false);
+            }
+
+            // submission projects are special here. The fields generated inside the Script object is private, but
+            // further submissions can bind to them.
+            await AddSubmissionDependentProjectsAsync(solution, symbolOrigination.sourceProject, dependentProjects, cancellationToken).ConfigureAwait(false);
+
+            return [.. dependentProjects];
+        }
     }
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs:193]: >} [](http://example.com/codeflow?start=4&length=1)

nit: maybe pool projectIdsToReferencingSubmissionIds below too?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs:193]: done.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs]: Maybe codeflow is tripping on me, but I'm not seeing this changed

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs]: oops.  in now.

**Pull request merged**